### PR TITLE
Rewrite $E.setState for Single ID Parameter

### PR DESCRIPTION
**Pull request by noi5e**
Related to #9004, OOP refactor.

Prerequisite: Merge #9104 first!

I think I'm getting the hang of `git rebase`. The merge conflicts won't be so bad here (I hope) once #9104 is merged. Also playing around with `git squash`ing all of the commits from `git rebase` into one commit, so the PR history is cleaner.

Still figuring out a lot of things with Git, pointers welcome!

This PR ended up being a lot bigger than I originally intended... There were a lot of interlocking changes. Sorry! I hope it will become easier to read once #9104 is merged beforehand. ("Files changed" is updated live, right?)

### `Renamed #text-input to #text-input-main`
I did this to standardize my ID naming conventions with the other comment elements:

- **REPLY**:
  - `text-input-reply-123`
  - `comment-preview-reply-123`
- **EDIT**:
  - `text-input-edit-123`
  - `comment-preview-edit-123`
- **MAIN**:
  - `comment-preview-main`
  - `textarea` for main:
    - **currently:** `#text-input`
    - **after this PR:** `#text-input-main`

### `Rewrite $E.setState() for Single ID Parameter`
With the above change, it's easier to `setState(single ID)`, rather than what's happening currently: `setState(textarea ID, preview ID)`.

Basically in a nutshell, since the IDs are standardized now, when I `setState`, I can just grab the ID that's hanging off the end of `#text-input-{ID}` or `#comment-preview-{ID}`. 

Currently, there's some complicated logic to figure out whether or not the comment ID is `main` and switching to a blank string. That's too complicated in the long run.

---
(This issue is part of the larger Comment Editor Overhaul Project with Outreachy. Refer to Planning Issue #9069 for more context)

### CODE REVIEW

File: app/views/comments/_form.html.erb [hunk 1]
```diff
@@ -76,7 +75,7 @@
             <!-- http://stackoverflow.com/questions/11235206/twitter-bootstrap-form-file-element-upload-button -->
             <label for="fileinput-choose-one-<%= comment_form_id %>">
               <input id="fileinput-choose-one-<%= comment_form_id %>" type="file" name="image[photo]" style="display:none;" />
-              <a onClick="handleClick()" style="cursor: pointer;" class="d-none d-md-inline text-underline"><%= translation('comments._form.choose_one') %></a>
```
Comment (noi5e) [inline app/views/comments/_form.html.erb:79]: Deleting this function `handleClick`, similar to how I deleted `setInit`. Basically, I've decided that it's best not to `setState` when opening forms. For more info, check out [this write-up](https://github.com/publiclab/plots2/issues/9004#issue-784759413)

File: app/views/comments/_edit.html.erb [hunk 2]
```diff
@@ -39,117 +59,43 @@
         </p>
       </div>
     </div>
-  <script type="application/javascript">
-//   function setInit(id) {
```
Comment (noi5e) [inline app/views/comments/_edit.html.erb:43]: Deleting this code.

File: app/assets/javascripts/editor.js [hunk 3]
```diff
@@ -23,117 +25,107 @@ $E = {
         return code;
       }
     });
-  },
-  setState: function(textarea = 'text-input', preview = 'comment-preview-main', title = 'title') {
-    $E.title = $('#' + title + 'title'); // not sure why this exists? seems like $E.title is always #title
-    $E.textarea = $('#' + textarea);
+  }
+  setState(commentFormID = "main") {
+    $E.textarea = $("#text-input-" + commentFormID);
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:30]: `setState` now takes one parameter, the comment form ID. Before, it was taking `preview ID` and `textarea ID`.

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -66,19 +66,7 @@ def feature_node(title)
...
-      'main'
-    when :reply
-      'reply-' + reply_to.to_s
-    when :responses
-      'responses'
-    end
-  end
-
-  # used in views/editor/_toolbar.html.erb
-  def get_toolbar_element_id(location, reply_to, comment_id)
```
Comment (noi5e) [inline app/helpers/application_helper.rb:81]: Deleted this helper, which was very similar to `get_comment_form_id`.

File: app/assets/javascripts/editor.js [hunk 5]
```diff
@@ -23,131 +26,122 @@ $E = {
...
     // textarea
     $E.textarea = ($D.selected).find('textarea').eq(0);
     $E.textarea.bind('input propertychange',$E.save);
     // preview
     $E.preview = ($D.selected).find('.comment-preview').eq(0);
-  },
-  isRichTextEditor: function(url) {
-    // this RegEx matches three different cases where the legacy editor is still used:
+  }
+  isSingleFormPage(url) {
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:47]: Renaming this function for greater accuracy.

File: app/assets/javascripts/editor.js [hunk 6]
```diff
@@ -23,131 +26,122 @@ $E = {
...
+  // h6() {
+  //   $E.wrap('######','')
+  // }
+  // h7() {
+  //   $E.wrap('#######','')
+  // }
   // this function is dedicated to Don Blair https://github.com/donblair
-  save: function() {
-    localStorage.setItem('plots:lastpost',$E.textarea.val())
-    localStorage.setItem('plots:lasttitle',$E.title.val())
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:109]: Deleted the `lasttitle`, it seems to be unused. Will be refactoring this code anyway in the next few weeks.

File: app/assets/javascripts/editor.js [hunk 7]
```diff
@@ -23,131 +26,122 @@ $E = {
...
+  }
+  toggle_preview() {
     let previewBtn;
     let dropzone;
     // if the element is part of a multi-comment page,
     // ensure to grab the current element and not the other comment element.
-    previewBtn = $(this.textarea.context).find('.preview-btn');
-    dropzone = $(this.textarea.context).find('.dropzone');
+    previewBtn = $("#toggle-preview-button-" + this.commentFormID);
+    dropzone = $("#dropzone-large-" + this.commentFormID);
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:136]: Rewrote these assignments so it sets on a specific ID, instead of searching through the DOM.

File: app/assets/javascripts/editor.js [hunk 8]
```diff
@@ -1,13 +1,16 @@
...
-    // default parameters point toward either:
-    //   1. the comment form at the bottom of multi-comment wikis/questions/research notes
-    //   2. the only editor form on /wiki/new and /wiki/edit
-    $E.setState();
-    
+class Editor {
+  // default parameter references the ID of:
+  //   1. main comment form in multi-comment wikis, questions, & research notes.
+  //   2. the only editor form on /wiki/new and /wiki/edit
+  constructor(commentFormID = "main") {
```
Comment (Sagarpreet) [inline app/assets/javascripts/editor.js:5]: There will be merge conflicts after we merge this PR: https://github.com/publiclab/plots2/pull/9104/files

File: app/assets/javascripts/editor.js (hunk 3 above)
Comment (Sagarpreet) [reply app/assets/javascripts/editor.js:32]: AWESOME !! more reliable now 🎉

**Review by Sagarpreet — COMMENTED** [review]
Good approach 💯

File: app/assets/javascripts/editor.js (hunk 5 above)
Comment (jywarren) [reply app/assets/javascripts/editor.js]: nice

**Review by jywarren — APPROVED** [review]

**Comment by jywarren** [comment]
🎉 🎉 🎉 🎉
